### PR TITLE
Workaround unsupported freeze insn

### PR DESCRIPTION
**Pull request by LU-JOHN**
Workaround unsupported freeze insn by:
1. replacing uses of freeze's result with freeze's source or a random (but compilation reproducible) constant if freeze's source is undef/poison
2. deleting freeze insn.

Long term solution is to add a freeze instruction extension in SPIR-V.  Issue is tracked in (https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1140)

### CODE REVIEW

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp [hunk 1]
```diff
@@ -413,6 +413,14 @@ bool SPIRVRegularizeLLVMBase::regularize() {
           if (isa<PossiblyExactOperator>(BO) && BO->isExact())
             BO->setIsExact(false);
         }
+
+        // FIXME: This is not valid handling for freeze instruction
+        if (auto *FI = dyn_cast<FreezeInst>(&II)) {
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVRegularizeLLVM.cpp:418]: Please also check for undef/poison values and materialize them

**Comment by asudarsa** [comment]
Please mention the issue (https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1140) and the long term solution in your PR description.

Thanks

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp (hunk 1 above)
Comment (LU-JOHN) [reply lib/SPIRV/SPIRVRegularizeLLVM.cpp:418]: Updated PR to materialize frozen undef/poison with a random constant.

**Review by MrSidims — APPROVED** [review]

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp [hunk 2]
```diff
@@ -413,6 +413,21 @@ bool SPIRVRegularizeLLVMBase::regularize() {
...
+          auto *V = FI->getOperand(0);
+          if (isa<UndefValue>(V)) {
+            if (V->getType()->isIntegerTy())
+              V = ConstantInt::get(V->getType(), 0);
+            else if (V->getType()->isFPOrFPVectorTy())
+              V = ConstantFP::get(V->getType(), 0.0);
```
Comment (svenvh) [inline lib/SPIRV/SPIRVRegularizeLLVM.cpp:424]: Would it be better to use `Constant::getNullValue` here, which would cover any type?

Comment (LU-JOHN) [reply lib/SPIRV/SPIRVRegularizeLLVM.cpp]: Updated to use Constant::getNullValue.  Added testing for ptr type which needs this change.

**Review by svenvh — APPROVED** [review]

**Comment by MrSidims** [comment]
Restarting CI

**Pull request merged**